// fhir-works-on-aws-interface/lib/errors/ResourceConflictError.js
"use strict";
/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isResourceConflictError = exports.ResourceConflictError = void 0;
// eslint-disable-next-line import/prefer-default-export
class ResourceConflictError extends Error {
    constructor(resourceType, id, message) {
        const msg = message || `Resource ${resourceType}/${id} is conflicting`;
        // Node Error class requires passing a string message to the parent class.
        super(msg);
        Object.setPrototypeOf(this, ResourceConflictError.prototype);
        this.resourceType = resourceType;
        this.id = id;
        this.name = this.constructor.name;
        this.isResourceConflictError = true;
    }
}
exports.ResourceConflictError = ResourceConflictError;
function isResourceConflictError(error) {
    return Boolean(error) && error.isResourceConflictError === true;
}
exports.isResourceConflictError = isResourceConflictError;
